bots/jep/src/main/java/org/openjdk/skara/bots/jep/JEPBot.java

```diff
@@ -0,0 +1,145 @@
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+
+package org.openjdk.skara.bots.jep;
+
+import org.openjdk.skara.bot.Bot;
+import org.openjdk.skara.bot.WorkItem;
+import org.openjdk.skara.forge.HostedRepository;
+import org.openjdk.skara.issuetracker.IssueProject;
+
+import java.nio.file.Path;
+import java.util.Collection;
+import java.util.List;
+import java.util.logging.Logger;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+public class JEPBot implements Bot, WorkItem {
+    final static String JEP_LABEL = "jep";
+    private final Logger log = Logger.getLogger("org.openjdk.skara.bots");
+    private final static Pattern jepMarkerPattern = Pattern.compile("<!-- jep: '(.*?)' '(.*?)' '(.*?)' -->");
+    private final HostedRepository repo;
+    private final IssueProject issueProject;
+
+    JEPBot(HostedRepository repo, IssueProject issueProject) {
+        this.repo = repo;
+        this.issueProject = issueProject;
+    }
+
+    @Override
+    public boolean concurrentWith(WorkItem other) {
+        if (!(other instanceof JEPBot otherBot)) {
+            return true;
+        }
+        return !repo.isSame(otherBot.repo);
+    }
+
+    @Override
+    public Collection<WorkItem> run(Path scratchPath) {
+        var prs = repo.pullRequests();
+        for (var pr : prs) {
+            var jepComment = pr.comments().stream()
+                    .filter(comment -> comment.author().equals(pr.repository().forge().currentUser()))
+                    .flatMap(comment -> comment.body().lines())
+                    .map(jepMarkerPattern::matcher)
+                    .filter(Matcher::find)
+                    .reduce((first, second) -> second)
+                    .orElse(null);
+            if (jepComment == null) {
+                if (pr.labelNames().contains(JEP_LABEL)) {
+                    pr.removeLabel(JEP_LABEL);
+                }
+                log.info("No jep command found in comment for PR-" + pr.id());
+                continue;
+            }
+
+            var issueId = jepComment.group(2);
+            if ("unneeded".equals(issueId)) {
+                if (pr.labelNames().contains(JEP_LABEL)) {
+                    pr.removeLabel(JEP_LABEL);
+                }
+                log.info("Found `/jep unneeded` command");
+                continue;
+            }
+
+            var issueOpt = issueProject.issue(issueId);
+            if (issueOpt.isEmpty()) {
+                if (pr.labelNames().contains(JEP_LABEL)) {
+                    pr.removeLabel(JEP_LABEL);
```

Comment:
If we remove the label in this situation, then the PR bot will mark the PR as `[x] Change requires a JEP request to be targeted`, which isn't true. I think the label should stay on. The jep command will not add the label unless it finds the JEP. I can imagine the following situations where this could happen:

* Jira is having a temporary outage - In this case we should do nothing.
* The JEP is no longer visible (e.g. made oracle confidential) - The user/author needs to decide how to handle this, by referencing another JEP or /jep unneeded.
* The JEP comment is corrupted - I don't think we can expect the bot to figure out this on its own.

Since this is an error state that shouldn't be possible to reach from normal operation, we should log SEVERE so that an admin gets notified. The same goes for the case below where the issue is of the wrong type.

Reply:
Fixed.